File: jihyuk/Programmers/전화번호 목록.cpp
/**
 1) 소요 시간 : 45분

 2) 사용 알고리즘 및 자료구조
	- map 자료구조 사용
 
 3) 풀면서 겪었던 문제
	- string 대소 비교시, string 크기 차이에 따른 리턴값을 헷갈림
	- 정렬 후 2중 for문을 사용해서 비효율적으로 구현. 추 후 1중 for문으로 다시 구현
	
 */

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool solution(vector<string> phone_book)
{
	/*
	 * 문자열 정렬
	 * 알파벳 순으로 정렬
	 * 접두사가 같고 길이만 다른 경우, 짧은 문자열이 앞쪽으로 배치됨
	 */
	sort(phone_book.begin(), phone_book.end());

	for (size_t i = 0; i < phone_book.size() - 1; ++i)
	{
		/*
		 * 알파벳순으로 정렬되어 있고, 동일한 접두사일 경우 짧은 문자열이 앞쪽에 배치됨
		 * 따라서 앞뒤로만 비교해주면 접두사 포함 관계를 판별할 수 있음
		 */

		// find() 함수로 찾았을 때 리턴 값 : 0
		if (0 == phone_book[i + 1].find(phone_book[i]))
		{
			return false;
		}
	}

	return true;
}

File: chelseashin/Programmers/01_Hash/전화번호목록.py
# phone_book = ["119", "97674223", "1195524421"]
# phone_book = ["123", "456", "789"]
phone_book = ["12", "123", "1235", "567", "88"]

# def solution(phone_book):
#     phone_book.sort(key=lambda x: len(x))
#     print(phone_book)
#     p = phone_book[0]
#     plen = len(p)
#     for num in phone_book[1:]:
#         if p == num[:plen]:
#             return False
#     return True

# def solution(phoneBook):
#     phoneBook = sorted(phoneBook)
#     print(phoneBook)
#     for p1, p2 in zip(phoneBook, phoneBook[1:]):
#         print(p1, p2)
#         if p2.startswith(p1):
#             return False
#     return True

def solution(phone_book):
    # hashmap = {}
    # for num in phone_book:
    #     hashmap[num] = 1
    # print(hashmap)
    for num in phone_book:
        temp = ""
        for n in num:
            temp += n
            if temp in phone_book and temp != num:
                return False
    return True

print(solution(phone_book))
